src/som/interpreter/actors/ReceivedRootNode.java

```diff
@@ -20,33 +21,44 @@ protected ReceivedRootNode(final Class<? extends TruffleLanguage<?>> language,
     assert sourceSection != null;
   }
 
-  protected final void resolvePromise(final VirtualFrame frame, final SResolver resolver, final Object result) {
+  protected final void resolvePromise(final VirtualFrame frame, final SResolver resolver, final Object result, final boolean isBreakpointOnResolutionAtSender, final boolean isBreakpointOnResolutionAtRcvr) {
+    // lazy initialization of resolution node
     if (resolve == null) {
+      CompilerDirectives.transferToInterpreterAndInvalidate();
       if (resolver == null) {
         this.resolve = insert(new NullResolver(getSourceSection()));
       } else {
-        this.resolve = insert(ResolvePromiseNodeFactory.create(false, getSourceSection(), null, null));
+        this.resolve = insert(ResolvePromiseNodeFactory.create(false, getSourceSection(), null, null, null, null));
       }
     }
 
-    resolve.executeEvaluated(frame, resolver, result);
+    // resolve promise
+    resolve.executeEvaluated(frame, resolver, result, isBreakpointOnResolutionAtSender, isBreakpointOnResolutionAtRcvr);
   }
 
+  /**
+   * Promise resolver for the case that the actual promise has been optimized out.
+   */
   public final class NullResolver extends ResolvePromiseNode {
-
     public NullResolver(final SourceSection source) {
       super(false, source);
     }
 
     @Override
-    public Object executeEvaluated(final VirtualFrame frame, final SResolver receiver, final Object argument) {
+    public Object executeEvaluated(final VirtualFrame frame, final SResolver receiver, final Object argument, final boolean isBreakpointOnResolutionAtSender, final boolean isBreakpointOnResolutionAtRcvr) {
       assert receiver == null;
+      if (isBreakpointOnResolutionAtSender) {
+        haltNode.executeEvaluated(frame, argument);
+      }
+      if (isBreakpointOnResolutionAtRcvr) {
+        //TODO: add a tag to the send node to disable or remove the button
+      }
       return null;
     }
 
     @Override
     public Object executeEvaluated(final VirtualFrame frame, final Object receiver,
-        final Object argument) {
+        final Object argument, final Object isBreakpointOnResolutionAtSender, final Object isBreakpointOnResolutionAtRcvr) {
```

Comment:
yes, but please use standard style comments, not javadoc style comments. So either `/* */` or `//` per line.